src/yahtzee_clj/core.clj

```diff
@@ -9,11 +10,80 @@
   "Rolls a hand of 5 six sided dice."
   (repeatedly 5 roll))
 
-(defn score-singles [hand]
-  "Scores the singles."
+(defn score-chance [hand]
+  "Scores the chance - i.e. sum of hand."
   (reduce + hand))
 
+(defn score-singles [hand single]
+  "Scores the singles"
+  (reduce + (filter #(= single %) hand))
+  )
+
+(defn score-of-a-kind [hand kind-type]
+  "Scores three, four and five of a kind"
+  (let [most-common (last (sort-by val (frequencies hand)))
+        face (first most-common)
+        count (last most-common)]
+    (cond
+      (< count 3) 0
+      (and (>= count 3) (= kind-type 3)) (* face 3)
+      (and (>= count 4) (= kind-type 4)) (* face 4)
+      (and (= count 5) (= kind-type 5)) 50
+      :else 0
+      )
+    ))
+
+;; TODO: this is horrible!
+(defn score-full-house [hand]
+  "Scores three, four and five of a kind"
+  (let [freqs (sort-by val (frequencies hand))
+        most-common (last (last freqs))
+        second-most-common (last (last (butlast freqs)))]
+    (if (and
+          (= most-common 3)
+          (= second-most-common 2))
+      25 0)))
+
+;; TODO: use threading macro to make it easier to read?
+(defn hand-to-sorted-string [hand]
+  "Returns a hand into a sorted string having removed dupes"
+  (join "" (sort (into [] (set (map str hand)))))
+  )
+
+(defn score-straight [hand straight_length]
+  "Scores small and large straight"
+  (let [valid_small '("1234" "2345" "3456")
```

Comment:
Perhaps we should use some higher order functions here to permanently tie valid small to the score of 30 and the possible straights?

(defn score-straight
  [valid-string-hands score]
  (fn [hand]
    (let [sorted-string-hand (hand-to-sorted-string hand)]
      (if (some #(includes? sorted-string-hand %) valid-string-hands)
        score
        0))))

(def score-low-straight (score-straight '("1234" "2345" "3456") 30))